### PR TITLE
Adding rule to prevent @action use on async methods

### PR DESCRIPTION
**Pull request by thejuan**
@action decorator shouldn't be used with async methods, flow should be used instead.

### CODE REVIEW

**Comment by eranshabi** [comment]
Hey @thejuan , thanks for the contribution!
I read [here](https://mobx.js.org/best/actions.html) that if the async @action function awaits a functions that has the @action decorator, then things will work, but I see that your rule will prevent this.
Is this intended? Can you please support this use case too?

**Comment by thejuan** [comment]
Hi,

That link is broken for me. But you can read about it here
https://mobx.js.org/best/actions.html

**Comment by eranshabi** [comment]
it's the same link:

> if you have a setTimeout, promise.then or async construction, and in that callback some more state is changed, those callbacks should be wrapped in action as well!

**Comment by thejuan** [comment]
A function like this is invalid

```
@action
public async myMethod(){
   const data = await this.getData();
   this.observableValue = data.value; // This needs to be in an action   
}
```

It needs to be written like this

```
@action
public async myMethod(){
   const data = await this.getData();
   runInAction(() => this.observableValue = data.value);
}
```

Or it needs to be written using flow.
For async methods we use flow, and use this rule so people don't accidentally implement the first option which doesn't work.

**Comment by thejuan** [comment]
The rule shouldn't be on by default for this package I wouldn't think, but its useful for codebases that don't use `runInAction`

File: src/rules/mobxReactNoAsyncActionRule.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*"no-async-with-action": context => ({
```
Comment (eranshabi) [inline src/rules/mobxReactNoAsyncActionRule.ts]: please remove commented out code

Comment (thejuan) [reply src/rules/mobxReactNoAsyncActionRule.ts]: Oops, forgot about that! Done.

**Comment by eranshabi** [comment]
Published

**Pull request merged**